### ISSUE TITLE
Drop PJAX

### ISSUE THREAD

**Issue by samdark** [yiisoft/yii2#15383]
- [ ] Remove PJAX itself.
- [ ] Remove PJAX-specific headers usage.

**Comment by Renkas** [comment]
Any alternatives? Should implement this myself if I want to use in 2.1? Move it to an extension maybe?

**Comment by samdark** [comment]
You can use the library directly. If you want to convert 2.0 PJAX into extension and support it, we'll link to it from the UPGRADE.

**Comment by bscheshirwork** [comment]
what the diff between clear PJAX and yii2-pjax?

buildin widgets will be work with clear pjax directly? like JQuery?

**Comment by samdark** [comment]
There will be no buildin widgets for PJAX.

**Comment by sniper888** [comment]
Why did that make a decision?

**Comment by samdark** [comment]
1. Too many problems with it. We have to patch the library.
2. It's not what current JavaScript trends are.
3. Yii is a PHP framework so we should focus on PHP and PJAX takes too much time to fix it.

**Comment by papppeter** [comment]
Hi,

could you maybe consider to drop it in favour of https://github.com/MoOx/pjax ?
so we still have this nice functionality.

**Comment by samdark** [comment]
Nope. Won't do it out of the box. It is decided. See above.

**Comment by klimov-paul** [comment]
Resolved by commit 603c084

**Comment by githubjeka** [comment]
Check checkboxes above ;)

**Comment by schmunk42** [comment]
For the record, there's also https://github.com/turbolinks/turbolinks - might be an alternative for some PJAX features

**Comment by SilverFire** [comment]
No more mighty JS libs, please :)

**Comment by FabrizioCaldarelli** [comment]
I think that a modern php framework should provide packages and libraries to speed up also ui development (it has been done adopting bootstrap). So, I can be agree to remove pjax implementation from core, but changes regard also framework implementation  (see pjax request handling insideYii). Although I could create an extension to replace currently pjax implementation (packaging current extension inside another extension), will that continue to work also with changes made inside Yii?

**Comment by bscheshirwork** [comment]
@FabrizioCaldarelli 
> (packaging current extension inside another extension)

another extension is a pure pjax?

**Comment by FabrizioCaldarelli** [comment]
My question is:could it possible to keep handling pjax request inside the code, so the only thing to do is creating an extension to use on Yii?  I have a lot of pjax code and It will break after upgrading to 2.1,without a fast solution.

**Comment by bscheshirwork** [comment]
+1 
We need a external tool same at current in-box tool; This is important.

**Comment by samdark** [comment]
No. We won't keep it. It it entirely possible to handle standard request-response in PJAX-manner w/o any extra modifications to the framework itself.

**Comment by rugabarbo** [comment]
If you need it you can create extension for 2.1 and share it using github.

**Comment by FabrizioCaldarelli** [comment]
@samdark I see that in 2.1 it has been removed much code in Request that recognize pjax Request. So if I create a new extension, Yii will not distinguish get/post request from pjax request. Is it right?

**Comment by samdark** [comment]
Yes. That is correct.

**Comment by FabrizioCaldarelli** [comment]
So should the new extension apply a path to Yii code adding again the same code that has been removed? Is it possibile  to keep at least Request code in Yii? Otherwise we have to rewrite the same code

**Comment by FabrizioCaldarelli** [comment]
My curiosity. Removing Pjax we lose real time interaction, coming back to full page reload (90s years style). What should be the modern js approach to give a real time appeal to frontend ui? For example, how grid view refresh will work without pjax? Full page reload?

**Comment by schmunk42** [comment]
The vast majority of the removed code belonged to the Pjax widget and asset. Those can be replaced with an extension.

For mimicking `Request::getIsPjax()`  a behavior could help.

**Comment by FabrizioCaldarelli** [comment]
I'm agree to keep as clear as possible Yii core, but I'm thinking at all index.php files that contain GridView wrapped by PJax by default. I've many projects that by default have GridView inside a Pjax, to take advantage of updating of a single part of view.
So switching to 2.1 version that files will not work anymore, and I think that many other people will be in this situation.
Can we think to a softer approach? Also to keep native app appeal that avoid full refresh of page when gridview update (filtering, changing page, etc.. etc..).

**Comment by FabrizioCaldarelli** [comment]
@samdark How will GridView work by default, after removing Pjax? Full page reload? Alternatives?

**Comment by schmunk42** [comment]
> So switching to 2.1 version that files will not work anymore, and I think that many other people will be in this situation.

If there'd be an extension `yii2-pjax` which would give you the Pjax widgets (with an aliased namespace?), what would not work in your opinion?

@yiisoft/core-developers What's about creating a repo in https://github.com/yiisoft-contrib for `yii2-pjax`?

**Comment by FabrizioCaldarelli** [comment]
@schmunk42 I think that we need to keep handling pjax inside Request class, to avoid to add a behaviour or a change to Request flow.
If we keep that, then we can split Pjax in an extension, keeping the same namespace if possible. So switching will be completely transparent.

**Comment by FabrizioCaldarelli** [comment]
@schmunk42 But until now I can figure how replace Pjax UI behaviour with some other alternative. How do you think about it? How will you reproduce same appeal of Pjax (for example, refreshing gridview), without a full page refresh?

**Comment by schmunk42** [comment]
@FabrizioCaldarelli This is the PJAX part of `Request` - https://github.com/yiisoft/yii2/blob/2.0.14/framework/web/Request.php#L465-L472

What else do you think of?

**Comment by FabrizioCaldarelli** [comment]
@schmunk42 Also in this:
https://github.com/yiisoft/yii2/blob/2.0.14/framework/web/Response.php#L854-L857

**Comment by schmunk42** [comment]
Maybe the Request/Response code could stay, there is also handling of edge cases like `X-Ie-Redirect-Compatibility`. Having those checks and features would not hurt - for around 10 lines of code, IMO.

**Comment by FabrizioCaldarelli** [comment]
Exactly, I think that leaving that code will not impact anything.

**Comment by schmunk42** [comment]
> No. We won't keep it. It it entirely possible to handle standard request-response in PJAX-manner w/o any extra modifications to the framework itself.

@samdark Keeping the code or do you have an idea how to do the Response stuff?

**Comment by bscheshirwork** [comment]
>  we need to keep handling pjax inside Request class

may be we need use closure for it?

**Comment by samdark** [comment]
X-Redirect is still there since it's common. PJAX-specific headers handling won't be in the core since it's that library only specific header.

**Comment by schmunk42** [comment]
But is `X-Ie-Redirect-Compatibility` not highly specific for Internet Explorer?

Will be or is threre a way to post-process the Response in Yii 2.1?

**Comment by samdark** [comment]
Maybe.

**Comment by FabrizioCaldarelli** [comment]
@samdark so gridview is planned to use full page reload to filter and to navigate the pages?

**Comment by samdark** [comment]
Out of the box — yes.

**Comment by FabrizioCaldarelli** [comment]
In a previous comment you refer that pjax does not follow js trends. What are last js trends?

**Comment by samdark** [comment]
Something like MVVM. https://vuejs.org/, https://reactjs.org/, https://angular.io/.

**Comment by schmunk42** [comment]
> What are last js trends?

Developing a new trendy framework every half-a-year or so :speak_no_evil:, which makes it pretty hard to follow the latest trends :stuck_out_tongue_winking_eye:

**Comment by samdark** [comment]
Yes, that :) Everything's changing too fast so it makes no sense to try to tightly integrate anything advanced from JS world. Basic things are OK though.

**Comment by FabrizioCaldarelli** [comment]
@samdark It could be good to create an official yii2-pjax repository to put all deleted data from core. So the extension would refer officially to Yii2, instead using personal repository.
If you think that it is wrong, I could create an extension for this. 
Let me know.

**Comment by SilverFire** [comment]
Would you like to support this repo?

**Comment by FabrizioCaldarelli** [comment]
@SilverFire Yes. I think that Yii should continue to support Pjax.

**Comment by samdark** [comment]
Core team won't support it for sure. If you want to take care of it, please create an extension and we'll link to it from upgrade notes.

**Comment by FabrizioCaldarelli** [comment]
Ok, it is clear. Because I understood that it is planned to split ui and js libraries from core, do we want to define a base namespace for all these (and future)  extensions? Could it be yii/ui/,,, or similar? Instead creating all extensions with owner and different namespace.

**Comment by samdark** [comment]
Different namespace is fine. Since it won't be yiisoft-maintained library, it's not a good idea for its namespace to be `\yii\*`.

**Comment by albertborsos** [comment]
I have just made an extension to the pjax widgets.
https://github.com/albertborsos/yii2-pjax

**Comment by bscheshirwork** [comment]
So... Also we can use origin pjax https://github.com/defunkt/jquery-pjax;

This is `asset` for connect it
https://github.com/bscheshirwork/yii2-jquery-pjax-asset

**Comment by samdark** [comment]
GitHub switched to https://github.com/martndemus/pjax btw.

**Comment by bscheshirwork** [comment]
@samdark
I don't understand what you mean...
GitHub community use https://github.com/martndemus/pjax 
or some of my links is broken?

**Comment by samdark** [comment]
Nope. Just mentioned it that GitHub switched from jQuery to native JS for PJAX.

**Comment by bscheshirwork** [comment]
May be "GitHub" as "engine of site https://github.com"? This cool...

**Comment by omzy83** [comment]
Why can't it just be made available as an official Pjax / client side JS library that we include in AppAsset.php, similar to how we do with BootstrapAsset?

**Comment by samdark** [comment]
That's what https://github.com/albertborsos/yii2-pjax is for.

**Comment by omzy83** [comment]
Yeah but that's now a third party extension. Different to a core extension maintained by the Yii dev team. 

One of the best things that separated Yii from other frameworks was the built-in client-side features, such as PJAX, client-side validation, Bootstrap integration, etc. Now that most of these are being stripped out, what USPs does Yii offer over other frameworks?

I know these features will still be available by installing the relevant package, etc, but surely a better option would be to allow these features to be enabled by config?

**Comment by samdark** [comment]
These features will be enable-able via config. Composer config is still config.

**Comment by omzy83** [comment]
That's fine - but personally I think there should be a library of "official" Yii extensions that are maintained by the Yii dev team, i.e. owned by "yiisoft" - and not third party developers. There are three reasons for this - firstly people have greater trust in official extensions, secondly it looks more professional, and thirdly there's no guarantee a third party extension will be maintained.

Then as long as you ensure there is a chapter for these extensions in the main documentation, it's all good.

**Comment by omzy83** [comment]
@samdark I understand your reasons for dropping PJAX support - but again I ask the question - after stripping out all of these features, what USPs does Yii offer over other frameworks?

Personally I have used Yii since version 1 - and I will continue using it. However how do you plan to make Yii appeal to developers currently using other frameworks? I've mentioned Laravel a few times in other posts and the reality is Yii has lost a lot of ground to Laravel, and stripping out features isn't going to help make up any ground.

**Comment by simialbi** [comment]
Maybe it would be a good idea to found some kind of organization maintaining yii-extension.
There could willing third party developers be members but all changes will be checked by yii team.

**Comment by samdark** [comment]
> what USPs does Yii offer over other frameworks

1. Performance.
2. Gii.
3. AR and DB layers.
4. Documentation (especially phpdoc).
5. i18n
6. ...

**Comment by samdark** [comment]
> but all changes will be checked by yii team

That's what we're trying to avoid to get more time for the core framework.
